storage/rocksdb/ha_rocksdb.cc

```diff
@@ -9646,6 +9676,16 @@ int ha_rocksdb::update_write_row(const uchar *const old_data,
     row_info.tx->incr_insert_count();
   }
 
+  ulonglong autoinc;
+  calc_autoinc_difference(old_data, new_data, table, autoinc);
+  if (autoinc) {
+    /*
+     A value for an AUTO_INCREMENT column
+     was specified in the UPDATE statement.
+    */
+    update_auto_incr_val_from_field();
```

Comment:
We already call `update_auto_incr_val_from_field` from `update_pk`, but it's not working as expected because it's using the 'old' row instead of the new one. I think a better fix is to unify these paths.

Reply:
Can we merge these two paths, since I think they're trying to do the same thing?

Reply:
Also, `table->next_number_field` isn't nullptr for me, but it is using the old row, which is why this bug reproduces for me.

```
(gdb) list
9395          bytes_written = row_info.old_pk_slice.size();
9396        }
9397      }
9398
9399      if (table->next_number_field) {
9400        update_auto_incr_val_from_field();
9401      }
9402
9403      int rc = HA_EXIT_SUCCESS;
9404      rocksdb::Slice value_slice;
(gdb) p table->next_number_field
$2 = (Field *) 0x7f2a9c825150
```

Reply:
That's weird. Can you set a memory watchpoint to show the call stack that sets table->next_number_field?
<pre>
Breakpoint 13, mysql_update (thd=0x2c8f850, table_list=0x7fffdc005710, fields=..., values=..., conds=0x0, order_num=0, order=0x0, limit=18446744073709551615, handle_duplicates=DUP_ERROR, 
    ignore=false, found_return=0x7ffff7e8d4a0, updated_return=0x7ffff7e8d498) at /workspace/mysql/sql/sql_update.cc:266
266	  if (!table_list->updatable)
(gdb) list
261	  }
262	
263	  THD_STAGE_INFO(thd, stage_init);
264	  table= table_list->table;
265	
266	  if (!table_list->updatable)
267	  {
268	    my_error(ER_NON_UPDATABLE_TABLE, MYF(0), table_list->alias, "UPDATE");
269	    DBUG_RETURN(1);
270	  }
(gdb) p table
$53 = (TABLE *) 0x7fffdc021180
(gdb) watch ((TABLE *) 0x7fffdc021180)->next_number_field
Hardware watchpoint 14: ((TABLE *) 0x7fffdc021180)->next_number_field
</pre>

It's not clear to me how to merge these two paths. Can we simply do below?
<pre>
--- a/storage/rocksdb/ha_rocksdb.cc
+++ b/storage/rocksdb/ha_rocksdb.cc
@@ -9410,7 +9410,7 @@ int ha_rocksdb::update_pk(const Rdb_key_def &kd,
     }
   }
 
-  if (table->next_number_field) {
+  if (table->found_next_number_field) {
     update_auto_incr_val_from_field();
   }
</pre>

Reply:
Checking `found_next_number_field` will probably be good enough for the fix. You might need to change `update_auto_incr_val_from_field` to look at new row.

For some reason, I'm seeing nullptr now for `next_number_field`, so I'm not sure how I got the previous results in the first place...

Reply:
Nice. I'll update the patch and also add a mtr test.